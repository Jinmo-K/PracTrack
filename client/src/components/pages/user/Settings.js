import React, { useState, useEffect } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
// Components
import TextField from '@material-ui/core/TextField';
import Collapse from '@material-ui/core/Collapse';
import Fade from '@material-ui/core/Fade';
// Functions
import { useInput } from '../../../hooks/useInput';
import { updateUser, resetAuthErrors } from '../../../actions/authActions';


/** 
 *   Page where users can edit account related details
 */
const Settings = ({ user, errors, updateUser, resetAuthErrors, updateUserStatus }) => {
  const {value:name, bindProps:bindName, setValue:setName} = useInput(user.name);
  const {value:email, bindProps:bindEmail, setValue:setEmail} = useInput(user.email);
  const {value:currPassword, bindProps:bindCurrPw, setValue: setCurrPw} = useInput('');
  const {value:password, bindProps:bindNewPw, setValue:setPw} = useInput('');
  const {value:password2, bindProps:bindNewPw2, setValue:setPw2} = useInput('');
  const [isChangingPw, setIsChangingPw] = useState(false);
  const [dirtyFlag, setDirtyFlag] = useState(false);

  const onSubmit = (e) => {
    e.preventDefault();
    let updateValues = { currPassword };
    if (name !== user.name)   updateValues.name = name;
    if (email !== user.email) updateValues.email = email;
    if (password) updateValues.password = password;
    if (password2) updateValues.password2 = password2;

    updateUser(user.id, updateValues)
  }

  /** For hiding/showing new password fields  */
  const toggleChangePw = () => {
    if (isChangingPw) {
      setPw('');
      setPw2('');
    }
    setIsChangingPw(!isChangingPw);
  }

  // Check for changed values to enable submit button
  useEffect(() => {
    if ((name !== user.name) || (email !== user.email) || (password && password2)) {
      setDirtyFlag(true);
    }
    else {
      setDirtyFlag(false);
    }
  }, [user, name, email, password, password2])

  // Resets the form using newly updated values. On unmount, resets all auth-related errors
  useEffect(() => {
    if (updateUserStatus === 'SUCCESS') {
      setName(user.name);
      setEmail(user.email);
      setCurrPw('');
      setPw('');
      setPw2('');
      setIsChangingPw(false);
    }
    return () => {
      if (Object.keys(errors).length) {
        resetAuthErrors(['name', 'email', 'currPassword', 'password', 'password2'])
      }
    }
  }, [user, updateUserStatus, errors, resetAuthErrors, setCurrPw, setEmail, setName, setPw, setPw2])

  return (
    <div className='row mt-3 justify-content-center login mx-auto align-items-center'>
      <form className='col-sm-10 col-md-7 col-lg-5 p-2 p-sm-5 mt-sm-2' noValidate onSubmit={onSubmit}>
        <div className='form-header'>
          <h2>User settings</h2>
        </div>

        {/* Name */}
        <div className='form-group'>
          <TextField 
            id='name'
            fullWidth
            label='Name'
            variant='outlined'
            error={!!errors.name}
            {...bindName} 
          />
          <span id='name-error' className="error-text">
            {errors.name}
          </span>
        </div>

        {/* Email */}
        <div className='form-group'>
          <TextField 
            id='email'
            fullWidth
            label='Email'
            variant='outlined'
            type='email'
            error={!!errors.email}
            {...bindEmail} 
          />
          <span id='email-error' className="error-text">
            {errors.email}
          </span>
        </div>

        {/* Current password */}
        <div className='form-group'>
          <TextField 
            id='pw'
            fullWidth
            label='Password'
            variant='outlined'
            type='password'
            error={!!errors.currPassword}
            {...bindCurrPw} 
          />
          <span id='pw-error' className="error-text">
            {errors.currPassword}
          </span>
        </div>

        {/* Change password button */}
        <button onClick={toggleChangePw} 
                id='change-pw-btn'
                type='button'
                className="btn btn-link pt-0 mb-2 forgot"
                style={{fontSize: '16px'}}
        >
          <u>{isChangingPw ? 'Cancel' : 'Change password'}</u>
        </button>

        <Collapse id='collapse' in={isChangingPw} collapsedHeight={0} timeout={200}>
          <Fade id='fade' in={isChangingPw} timeout={400}>
            <div>
              {/* New password */}
              <div className='form-group'>
                <TextField 
                  id='new-pw'
                  fullWidth
                  label='New password'
                  variant='outlined'
                  type='password'
                  error={!!errors.password}
                  {...bindNewPw} 
                />
                <span id='new-pw-error' className="error-text">
                  {errors.password}
                </span>
              </div>
              {/* Confirm new password */}
              <div className='form-group'>
                <TextField 
                  id='confirm-pw'
                  fullWidth
                  label='Confirm new password'
                  variant='outlined'
                  type='password'
                  error={!!errors.password2}
                  {...bindNewPw2} 
                />
                <span id='confirm-pw-error' className="error-text">
                  {errors.password2}
                </span>
              </div>
            </div>
          </Fade>
        </Collapse>

        <div className='form-group'>
          <button 
            className='btn btn-block' 
            type='submit' 
            disabled={!dirtyFlag}
            style={{fontWeight: 500}}
          >
            Save
          </button>
        </div>
      </form>
    </div>
  );
};


Settings.propTypes = {
  /** User object held in auth store state */
  user: PropTypes.object.isRequired,
  /** Auth-related errors store state */
  errors: PropTypes.object.isRequired,
  /** Dispatch function for updating a user with the given values */
  updateUser: PropTypes.func.isRequired,
  /** Dispatches action to reset auth errors */
  resetAuthErrors: PropTypes.func.isRequired,
  /** Indicates whether updating user has succeeded */
  updateUserStatus: PropTypes.string.isRequired,
}

const mapStateToProps = (state) => ({
  user: state.auth.user,
  errors: state.errors,
  updateUserStatus: state.status.updateUser,
});

const mapDispatchToProps = {
  updateUser,
  resetAuthErrors,
}

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(Settings);

export { Settings };