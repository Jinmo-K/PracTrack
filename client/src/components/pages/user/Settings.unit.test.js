import React from 'react';
import { shallow, mount } from 'enzyme';
import renderer from 'react-test-renderer';
import { Settings } from './Settings';

describe('<Settings />', () => {
  let wrapper;
  let props;

  beforeEach(() => {
    props = {
      user: {
        id: '0',
        name: 'Test User',
        email: 'test@test.com',
      },
      errors: {},
      updateUser: jest.fn(),
      resetAuthErrors: jest.fn(),
      updateUserStatus: ''
    };
    jest.clearAllMocks();
    wrapper = shallow(<Settings {...props} />);
  });

  it('Should render initial layout given required props', () => {
    let tree = renderer.create(<Settings {...props} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  describe('Name field', () => {
    it('Should reflect user input in name field', () => {
      wrapper.find('#name').simulate('change', { target: { value: 'test' } });
      expect(wrapper.find('#name').props().value).toEqual('test');
    });
  
    it('Should display name error', () => {
      props.errors = { name: 'Name error' };
      wrapper = shallow(<Settings {...props} />);
      expect(wrapper.find('#name-error').props().children).toEqual('Name error');
    });
  
    it('Should enable submit button when name field differs from original value', () => {
      wrapper = mount(<Settings {...props} />);
      wrapper.find('input#name').simulate('change', { target: { value: 'different' } });
      expect(wrapper.find('[type="submit"]').props().disabled).toEqual(false);
    });
  
    it('Should disable submit button when name field equals the original value', () => {
      wrapper = mount(<Settings {...props} />);
      wrapper.find('input#name').simulate('change', { target: { value: props.user.name } });
      expect(wrapper.find('[type="submit"]').props().disabled).toEqual(true);
    });
  });

  describe('Email field', () => {
    it('Should reflect user input in e-mail field', () => {
      wrapper.find('#email').simulate('change', { target: { value: 'test' } });
      expect(wrapper.find('#email').props().value).toEqual('test');
    });
  
    it('Should display e-mail error', () => {
      props.errors = { email: 'E-mail error' };
      wrapper = shallow(<Settings {...props} />);
      expect(wrapper.find('#email-error').props().children).toEqual('E-mail error');
    });
  
    it('Should enable submit button when e-mail field differs from original value', () => {
      wrapper = mount(<Settings {...props} />);
      wrapper.find('input#email').simulate('change', { target: { value: 'different' } });
      expect(wrapper.find('[type="submit"]').props().disabled).toEqual(false);
    });
  
    it('Should disable submit button when e-mail field equals the original value', () => {
      wrapper = mount(<Settings {...props} />);
      wrapper.find('input#email').simulate('change', { target: { value: props.user.email } });
      expect(wrapper.find('[type="submit"]').props().disabled).toEqual(true);
    });
  });

  describe('Password fields', () => {
    it('Should reflect user input in password field', () => {
      wrapper.find('#pw').simulate('change', { target: { value: 'test' } });
      expect(wrapper.find('#pw').props().value).toEqual('test');
    });
  
    it('Should display current password error', () => {
      props.errors = { currPassword: 'Password error' };
      wrapper = shallow(<Settings {...props} />);
      expect(wrapper.find('#pw-error').props().children).toEqual('Password error');
    });
  
    // New password field
    it('Should reflect user input in new password field', () => {
      wrapper.find('#new-pw').simulate('change', { target: { value: 'test' } });
      expect(wrapper.find('#new-pw').props().value).toEqual('test');
    });
  
    it('Should display new password error', () => {
      props.errors = { password: 'Password error' };
      wrapper = shallow(<Settings {...props} />);
      expect(wrapper.find('#new-pw-error').props().children).toEqual('Password error');
    });
  
    // Confirm password field
    it('Should reflect user input in confirm password field', () => {
      wrapper.find('#confirm-pw').simulate('change', { target: { value: 'test' } });
      expect(wrapper.find('#confirm-pw').props().value).toEqual('test');
    });
  
    it('Should display confirm password error', () => {
      props.errors = { password2: 'Password error' };
      wrapper = shallow(<Settings {...props} />);
      expect(wrapper.find('#confirm-pw-error').props().children).toEqual('Password error');
    });
  
    it('Should enable submit button when new and confirm password fields present', () => {
      wrapper = mount(<Settings {...props} />);
      wrapper.find('input#new-pw').simulate('change', { target: { value: 'test' } });
      wrapper.find('input#confirm-pw').simulate('change', { target: { value: 'test' } });
      expect(wrapper.find('[type="submit"]').props().disabled).toEqual(false);
    });
  
    it('Should disable submit button when either new or confirm password fields not present', () => {
      wrapper = mount(<Settings {...props} />);
      wrapper.find('input#new-pw').simulate('change', { target: { value:'test' } });
      expect(wrapper.find('[type="submit"]').props().disabled).toEqual(true);
      wrapper.find('input#new-pw').simulate('change', { target: { value:'' } });
      wrapper.find('input#confirm-pw').simulate('change', { target: { value:'test' } });
      expect(wrapper.find('[type="submit"]').props().disabled).toEqual(true);
    });
  
    it('Should reveal password fields when \'Change password\' button clicked', () => {
      wrapper.find('#change-pw-btn').simulate('click');
      expect(wrapper.find('#collapse').props().in).toEqual(true);
      expect(wrapper.find('#fade').props().in).toEqual(true);
    });
  
    it('Should hide password fields when \'Change password\' button clicked', () => {
      wrapper.find('#change-pw-btn').simulate('click');
      wrapper.find('#change-pw-btn').simulate('click');
      expect(wrapper.find('#collapse').props().in).toEqual(false);
      expect(wrapper.find('#fade').props().in).toEqual(false);
    });
  
    it('Should hide password fields when user has updated successfully', () => {
      wrapper = mount(<Settings {...props} />);
      wrapper.find('#change-pw-btn').simulate('click');
      wrapper.setProps({ updateUserStatus: 'SUCCESS' });
      wrapper.update();
      expect(wrapper.find('#collapse').first().props().in).toEqual(false);
      expect(wrapper.find('#fade').first().props().in).toEqual(false);
    });
  });

  describe('Cleanup', () => {
    it('Should dispatch resetAuthErrors on unmount if any error exists', () => {
      props.errors = { name: 'name error' };
      wrapper = mount(<Settings {...props} />);
      wrapper.unmount();
      expect(props.resetAuthErrors).toBeCalledTimes(1);
      expect(props.resetAuthErrors.mock.calls[0][0].includes('name')).toEqual(true);
      expect(props.resetAuthErrors.mock.calls[0][0].includes('email')).toEqual(true);
      expect(props.resetAuthErrors.mock.calls[0][0].includes('currPassword')).toEqual(true);
      expect(props.resetAuthErrors.mock.calls[0][0].includes('password')).toEqual(true);
      expect(props.resetAuthErrors.mock.calls[0][0].includes('password2')).toEqual(true);
    });
  
    it('Should not dispatch resetAuthErrors on unmount when there are no errors', () => {
      wrapper = mount(<Settings {...props} />);
      wrapper.unmount();
      expect(props.resetAuthErrors).toBeCalledTimes(0);
    });
  });

  describe('updateUser function', () => {
    const mockEvent = {
      preventDefault: jest.fn()
    };

    it('Should dispatch updateUser on submit', () => {
      wrapper.find('form').simulate('submit', mockEvent);
      expect(mockEvent.preventDefault).toBeCalledTimes(1);
      expect(props.updateUser).toBeCalledTimes(1);
    });

    it('Should dispatch updateUser with correct user ID', () => {
      wrapper.find('form').simulate('submit', mockEvent);
      expect(props.updateUser.mock.calls[0][0]).toEqual(props.user.id);
    });

    it('Should dispatch updateUser with current password value', () => {
      wrapper.find('#pw').simulate('change', { target: { value: 'newPw' } });
      wrapper.find('form').simulate('submit', mockEvent);
      expect(props.updateUser.mock.calls[0][1].currPassword).toEqual('newPw');
    });
  
    it('Should dispatch updateUser with new name value', () => {
      wrapper.find('#name').simulate('change', { target: { value: 'new' } });
      wrapper.find('form').simulate('submit', mockEvent);
      expect(props.updateUser.mock.calls[0][1].name).toEqual('new');
    });

    it('Should dispatch updateUser with new email value', () => {
      wrapper.find('#email').simulate('change', { target: { value: 'new' } });
      wrapper.find('form').simulate('submit', mockEvent);
      expect(props.updateUser.mock.calls[0][1].email).toEqual('new');
    });

    it('Should dispatch updateUser with new password value', () => {
      wrapper.find('#new-pw').simulate('change', { target: { value: 'new' } });
      wrapper.find('form').simulate('submit', mockEvent);
      expect(props.updateUser.mock.calls[0][1].password).toEqual('new');
    });

    it('Should dispatch updateUser with confirm password value', () => {
      wrapper.find('#confirm-pw').simulate('change', { target: { value: 'new' } });
      wrapper.find('form').simulate('submit', mockEvent);
      expect(props.updateUser.mock.calls[0][1].password2).toEqual('new');
    });
  });
});
