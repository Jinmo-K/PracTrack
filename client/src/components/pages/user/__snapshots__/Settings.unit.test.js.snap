// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Settings /> Should render initial layout given required props 1`] = `
<div
  className="row mt-3 justify-content-center login mx-auto align-items-center"
>
  <form
    className="col-sm-10 col-md-7 col-lg-5 p-2 p-sm-5 mt-sm-2"
    noValidate={true}
    onSubmit={[Function]}
  >
    <div
      className="form-header"
    >
      <h2>
        User settings
      </h2>
    </div>
    <div
      className="form-group"
    >
      <div
        className="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
      >
        <label
          className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined MuiFormLabel-filled"
          data-shrink={true}
          htmlFor="name"
          id="name-label"
        >
          Name
        </label>
        <div
          className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth MuiInputBase-formControl"
          onClick={[Function]}
        >
          <input
            aria-invalid={false}
            autoFocus={false}
            className="MuiInputBase-input MuiOutlinedInput-input"
            disabled={false}
            id="name"
            onAnimationStart={[Function]}
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            required={false}
            type="text"
            value="Test User"
          />
          <fieldset
            aria-hidden={true}
            className="PrivateNotchedOutline-root-59 MuiOutlinedInput-notchedOutline"
          >
            <legend
              className="PrivateNotchedOutline-legendLabelled-61 PrivateNotchedOutline-legendNotched-62"
            >
              <span>
                Name
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
      <span
        className="error-text"
        id="name-error"
      />
    </div>
    <div
      className="form-group"
    >
      <div
        className="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
      >
        <label
          className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined MuiFormLabel-filled"
          data-shrink={true}
          htmlFor="email"
          id="email-label"
        >
          Email
        </label>
        <div
          className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth MuiInputBase-formControl"
          onClick={[Function]}
        >
          <input
            aria-invalid={false}
            autoFocus={false}
            className="MuiInputBase-input MuiOutlinedInput-input"
            disabled={false}
            id="email"
            onAnimationStart={[Function]}
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            required={false}
            type="email"
            value="test@test.com"
          />
          <fieldset
            aria-hidden={true}
            className="PrivateNotchedOutline-root-59 MuiOutlinedInput-notchedOutline"
          >
            <legend
              className="PrivateNotchedOutline-legendLabelled-61 PrivateNotchedOutline-legendNotched-62"
            >
              <span>
                Email
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
      <span
        className="error-text"
        id="email-error"
      />
    </div>
    <div
      className="form-group"
    >
      <div
        className="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
      >
        <label
          className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined"
          data-shrink={false}
          htmlFor="pw"
          id="pw-label"
        >
          Password
        </label>
        <div
          className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth MuiInputBase-formControl"
          onClick={[Function]}
        >
          <input
            aria-invalid={false}
            autoFocus={false}
            className="MuiInputBase-input MuiOutlinedInput-input"
            disabled={false}
            id="pw"
            onAnimationStart={[Function]}
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            required={false}
            type="password"
            value=""
          />
          <fieldset
            aria-hidden={true}
            className="PrivateNotchedOutline-root-59 MuiOutlinedInput-notchedOutline"
          >
            <legend
              className="PrivateNotchedOutline-legendLabelled-61"
            >
              <span>
                Password
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
      <span
        className="error-text"
        id="pw-error"
      />
    </div>
    <button
      className="btn btn-link pt-0 mb-2 forgot"
      id="change-pw-btn"
      onClick={[Function]}
      style={
        Object {
          "fontSize": "16px",
        }
      }
      type="button"
    >
      <u>
        Change password
      </u>
    </button>
    <div
      className="MuiCollapse-container MuiCollapse-hidden"
      id="collapse"
      style={
        Object {
          "minHeight": "0px",
        }
      }
    >
      <div
        className="MuiCollapse-wrapper"
      >
        <div
          className="MuiCollapse-wrapperInner"
        >
          <div
            id="fade"
            style={
              Object {
                "opacity": 0,
                "visibility": "hidden",
              }
            }
          >
            <div
              className="form-group"
            >
              <div
                className="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
              >
                <label
                  className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined"
                  data-shrink={false}
                  htmlFor="new-pw"
                  id="new-pw-label"
                >
                  New password
                </label>
                <div
                  className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth MuiInputBase-formControl"
                  onClick={[Function]}
                >
                  <input
                    aria-invalid={false}
                    autoFocus={false}
                    className="MuiInputBase-input MuiOutlinedInput-input"
                    disabled={false}
                    id="new-pw"
                    onAnimationStart={[Function]}
                    onBlur={[Function]}
                    onChange={[Function]}
                    onFocus={[Function]}
                    required={false}
                    type="password"
                    value=""
                  />
                  <fieldset
                    aria-hidden={true}
                    className="PrivateNotchedOutline-root-59 MuiOutlinedInput-notchedOutline"
                  >
                    <legend
                      className="PrivateNotchedOutline-legendLabelled-61"
                    >
                      <span>
                        New password
                      </span>
                    </legend>
                  </fieldset>
                </div>
              </div>
              <span
                className="error-text"
                id="new-pw-error"
              />
            </div>
            <div
              className="form-group"
            >
              <div
                className="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
              >
                <label
                  className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined"
                  data-shrink={false}
                  htmlFor="confirm-pw"
                  id="confirm-pw-label"
                >
                  Confirm new password
                </label>
                <div
                  className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth MuiInputBase-formControl"
                  onClick={[Function]}
                >
                  <input
                    aria-invalid={false}
                    autoFocus={false}
                    className="MuiInputBase-input MuiOutlinedInput-input"
                    disabled={false}
                    id="confirm-pw"
                    onAnimationStart={[Function]}
                    onBlur={[Function]}
                    onChange={[Function]}
                    onFocus={[Function]}
                    required={false}
                    type="password"
                    value=""
                  />
                  <fieldset
                    aria-hidden={true}
                    className="PrivateNotchedOutline-root-59 MuiOutlinedInput-notchedOutline"
                  >
                    <legend
                      className="PrivateNotchedOutline-legendLabelled-61"
                    >
                      <span>
                        Confirm new password
                      </span>
                    </legend>
                  </fieldset>
                </div>
              </div>
              <span
                className="error-text"
                id="confirm-pw-error"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      className="form-group"
    >
      <button
        className="btn btn-block"
        disabled={true}
        style={
          Object {
            "fontWeight": 500,
          }
        }
        type="submit"
      >
        Save
      </button>
    </div>
  </form>
</div>
`;
